File: app/src/test/java/com/billpayments/features/settings/data/SettingsStoreTest.kt
package com.billpayments.features.settings.data

import com.billpayments.core.storage.InMemoryKeyValueStore
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.test.StandardTestDispatcher
import kotlinx.coroutines.test.TestScope
import kotlinx.coroutines.test.advanceUntilIdle
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Test

class SettingsStoreTest {

    @Test
    fun defaultsToSystemTheme() = runTest {
        val store = SettingsStore(StoredSettingsRepository(InMemoryKeyValueStore()), { "en" }, storeScope())
        assertEquals(AppThemeMode.System, store.state.value.themeMode)
    }

    @Test
    fun persistsAndRestoresTheme() = runTest {
        val keyValueStore = InMemoryKeyValueStore()
        val first = SettingsStore(StoredSettingsRepository(keyValueStore), { "en" }, storeScope())
        first.setThemeMode(AppThemeMode.Dark)
        advanceUntilIdle()
        val second = SettingsStore(StoredSettingsRepository(keyValueStore), { "en" }, storeScope())
        assertEquals(AppThemeMode.Dark, second.state.value.themeMode)
    }

    @Test
    fun languageFollowsSystemElseEnglish() = runTest {
        val arabicSystem = SettingsStore(StoredSettingsRepository(InMemoryKeyValueStore()), { "ar" }, storeScope())
        assertEquals("ar", arabicSystem.language.value)
        val frenchSystem = SettingsStore(StoredSettingsRepository(InMemoryKeyValueStore()), { "fr" }, storeScope())
        assertEquals("en", frenchSystem.language.value)
    }

    @Test
    fun refreshLanguagePicksUpLocaleChange() = runTest {
        var systemLanguage = "en"
        val store = SettingsStore(StoredSettingsRepository(InMemoryKeyValueStore()), { systemLanguage }, storeScope())
        assertEquals("en", store.language.value)
        systemLanguage = "hi"
        store.refreshLanguage()
        assertEquals("hi", store.language.value)
    }
}

private fun TestScope.storeScope(): CoroutineScope =
    CoroutineScope(SupervisorJob() + StandardTestDispatcher(testScheduler))

File: __tests__/features/settingsStore.test.ts
import { InMemoryKeyValueStore } from '@/core/storage/keyValueStore';
import { StoredSettingsRepository } from '@/features/settings/data/settingsRepository';
import { createSettingsStore } from '@/features/settings/data/settingsStore';

const make = () => {
  const store = new InMemoryKeyValueStore();
  return { store, settings: createSettingsStore({ settingsRepository: new StoredSettingsRepository(store) }) };
};

test('setLocale updates state and persists', () => {
  const { store, settings } = make();
  settings.getState().setLocale('hi');
  expect(settings.getState().settings.localeOverride).toBe('hi');
  expect(store.read('settings.locale')).toBe('hi');
});

test('setLocale null clears the persisted override', () => {
  const { store, settings } = make();
  settings.getState().setLocale('ar');
  settings.getState().setLocale(null);
  expect(settings.getState().settings.localeOverride).toBeNull();
  expect(store.read('settings.locale')).toBeNull();
});

test('setThemeMode updates state and persists', () => {
  const { store, settings } = make();
  settings.getState().setThemeMode('dark');
  expect(settings.getState().settings.themeMode).toBe('dark');
  expect(store.read('settings.themeMode')).toBe('dark');
});
